### PR TITLE
AO3-5765 Rails 5.2 upgrade

### PR DESCRIPTION
**Pull request by CristinaRO**
# Pull Request Checklist

* [x] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [x] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [x] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [x] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-5765
https://otwarchive.atlassian.net/browse/AO3-5766
https://otwarchive.atlassian.net/browse/AO3-5767
https://otwarchive.atlassian.net/browse/AO3-5768

## Purpose

Updated #3642.

## Testing Instructions

How can the Archive's QA team verify that this is working as you intended?

(If you have a Jira account with access, please update or comment on the issue
with any new or missing testing instructions.)

## References

Are there other relevant issues/pull requests/mailing list discussions?

## Credit

*What name do you want us to use to credit your work in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?*

(If you have a Jira account, please include the same name in the "Full name"
field on your Jira profile, so we can assign you the issues you're working on.)

*What pronouns do you prefer (she/her, he/him, zie/hir etc)?*

(If you do not fill in this section, we will use your GitHub account name and
they/them pronouns to avoid making assumptions about your gender.)

### CODE REVIEW

**Comment by CristinaRO** [comment]
Will obey the Hound later, wanted to celebrate getting a green Travis on the branch, and to get this out of my mental context and into the world.

File: Gemfile [hunk 1]
```diff
@@ -139,13 +137,13 @@ gem 'dalli'
 gem 'kgio', '2.10.0'
 
 group :test do
-  gem 'rspec', '~> 3.4'
-  gem 'rspec-rails', '~> 3.6.0'
+  gem 'rspec', '~> 3.8'
+  gem 'rspec-rails', '~> 3.8.2'
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply Gemfile:141]: Yeah, not going to change the whole Gemfile, but also not going to create a horrible mosaic of single and double quoted strings.

File: app/models/comment.rb [hunk 2]
```diff
@@ -65,19 +63,19 @@ def akismet_attributes
...
+  def moderate_update
+    return if !moderated_commenting_enabled? || is_creator_comment?
+    self.unreviewed = true if significant_content_change?
+  end
+
   def after_update
     users = []
     admins = []
 
-    if self.saved_change_to_edited_at? && self.saved_change_to_comment_content? && self.moderated_commenting_enabled? && !self.is_creator_comment?
```
Comment (tickinginstant) [inline app/models/comment.rb:73]: This comment from the original PR still hasn't been addressed: https://github.com/otwcode/otwarchive/pull/3642#discussion_r329283994

**Comment by tickinginstant** [comment]
Also this comment: https://github.com/otwcode/otwarchive/pull/3642#pullrequestreview-378891006

File: factories/comments.rb [hunk 3]
```diff
@@ -10,6 +10,14 @@
     pseud
   end
 
+  factory :guest_comment, class: Comment do
```
Comment (sarken) [inline factories/comments.rb:13]: hahaha sob, I typed up a comment asking why we added this when it's the same as the one above it... but now I see the one above it is some hybrid logged-in/guest comment monstrosity that should never occur in the wilds of the Archive, so that's cool and wonderful our comment tests have been using that. 🤦

Comment (tickinginstant) [reply factories/comments.rb:13]: On top of being simultaneously registered and non-registered, the comment factory defaults to having the commentable being a work, which is not a great model of what a real comment looks like, because [all comments are on chapters rather than on works](https://github.com/otwcode/otwarchive/blob/3b2cf664ea303b1066f7372b6930c9b51cef64b0/app/models/comment.rb#L367-L369).

(Incidentally, I suspect the work/chapter issue is why [this issue with comments being included in the stats](https://github.com/otwcode/otwarchive/blob/3b2cf664ea303b1066f7372b6930c9b51cef64b0/features/step_definitions/work_search_steps.rb#L137-L146) occurred. If the comments were generated with `FactoryBot.create(:comment, commentable: work)`, they wouldn't show up in the stats because the stats only include comments on chapters.)

Comment (sarken) [reply factories/comments.rb:13]: Our Slack emoji for sobbing blood would be handy right now.

I'll make an issue for fixing up the comment factory.

Comment (sarken) [reply factories/comments.rb:13]: [AO3-5999](https://otwarchive.atlassian.net/browse/AO3-5999) now exists for updating the existing factories, but it would be great if you could update the commentable on the new factory to be a chapter, @CristinaRO.

File: Gemfile [hunk 4]
```diff
@@ -73,7 +74,7 @@ gem 'elasticsearch', '6.8.0'
 gem 'aws-sdk'
 gem 'css_parser'
 
-gem 'cocaine'
+gem 'terrapin'
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply Gemfile:77]: Yeah, no, not unless we change the whole Gemfile in one go, or we land in conflict hell.

File: Gemfile [hunk 5]
```diff
@@ -37,7 +38,7 @@ gem 'rack-attack'
 
 # Version of redis-rb gem
 # We are currently running Redis 3.2.1 (7/2018)
-gem 'redis', ">=3.0"
+gem 'redis', "~> 3.3.5"
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply Gemfile:41]: The inconsistency in the same line hurts, though. 😬

File: app/models/comment.rb (hunk 2 above)
Comment (CristinaRO) [reply app/models/comment.rb:73]: Thanks for fishing that out of the well! (I find it very hard to follow the many comments on giant PRs, GitHub should really come up with a better way to wade through them...)

I don't quite understand what is the desired outcome, though - the return of the old conditionals but in the new method, like so:
```
    def significant_content_change?
      saved_change_to_edited_at? &&
        saved_change_to_comment_content? &&
        content_too_different?(comment_content, comment_content_was)
    end
```
or adding back the `edited_at` check, but replacing `saved_change_to_comment_content?` with `comment_content_changed?`?

> what the issue was: `self.comment_content_was` vs `self.comment_content_before_last_save`

I don't understand what this means.

File: Gemfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (redsummernight) [reply Gemfile:41]: You can change that now, as well as any lines that were edited anyway. We'll live with the inconsistency for a while.

[the remark this answers is not in the document]
Comment (redsummernight) [reply Gemfile:41]: There's some more old Rails stuff that can be removed:

https://github.com/otwcode/otwarchive/blob/057b4ff672ef4ef480b9b2fdddb710e1b052cba7/Gemfile#L15-L22

File: app/models/comment.rb [hunk 6]
```diff
@@ -212,6 +210,12 @@ def add_feedback_to_inbox(user)
       new_feedback.save
     end
 
+    def significant_content_change?
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/models/comment.rb:213]: Ah I see, the 4 instead of 2 has been there for a while, but it only noticed it for one method because that's the method we changed.

File: Gemfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (CristinaRO) [reply Gemfile:41]: done in the next push

**Comment by CristinaRO** [comment]
Could I beg indulgence from all the things that Hound finds? It's more Cerberus / Hydra than hound by now, I fix one thing, ten more pop up...

I'd rather have a PR that is "good enough" and moves, than a "perfect" one that keeps amassing conflicts and is postponed until the Doomsday after tomorrow.

File: app/models/comment.rb (hunk 2 above)
Comment (tickinginstant) [reply app/models/comment.rb:73]: I don't believe either of those fixes will work. Rails changed the behavior of `attribute_was` in 5.2, which is what [this deprecation warning that I copied from the 5.1 automated tests](https://github.com/otwcode/otwarchive/pull/3642#issuecomment-539723563) was about:
> DEPRECATION WARNING: The behavior of `attribute_was` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `attribute_before_last_save` instead. (called from after_update at /home/travis/build/tickinginstant/otwarchive/app/models/comment.rb:75)

The code from the original PR moves the check from `after_update` to `before_update`, which is why it uses `attribute_was` and `attribute_changed?` rather than `attribute_before_last_save` and `saved_change_to_attribute?`. But it sounds like you're proposing mixing them, which won't work regardless of when the check runs.

At this point, I think it'd be simplest to revert all of the changes to `comment.rb` (aside from the scope changes) and change the line `if content_too_different?(self.comment_content, self.comment_content_was)` to `if content_too_different?(comment_content, comment_content_before_last_save)`. That should also help prevent the Hound from barking about everything under the sun.

Comment (CristinaRO) [reply app/models/comment.rb:73]: Oh, I'm not proposing anything *g*, I was just trying to understand your original comment here https://github.com/otwcode/otwarchive/pull/3642#discussion_r329283994 and likely misinterpreting the `edited_at` bit.

(I didn't see the other comment because it wasn't directly on the code being discussed, and I never noticed the deprecation errors in the test suite. 😭 )

File: factories/comments.rb (hunk 3 above)
Comment (redsummernight) [reply factories/comments.rb:13]: #3870 added a factory for guest comments and was merged, you can use that now.

File: spec/controllers/admin/admin_users_controller_spec.rb [hunk 7]
```diff
@@ -92,24 +92,71 @@
...
+              roles: [role.id.to_s]
+            }
+          }
 
-        expect(user.roles.include?(role)).to be_truthy
+          it_redirects_to_with_notice(root_path, "User was successfully updated.")
+        end
+      end
+
+      %w[open_doors tag_wrangling].each do |admin_role|
```
[the remark this answers is not in the document]
Comment (sarken) [reply spec/controllers/admin/admin_users_controller_spec.rb:116]: No, we're updating this.

File: app/models/comment.rb (hunk 2 above)
Comment (CristinaRO) [reply app/models/comment.rb:73]: Well, I've done that, and then the feature test was failing. I've added some `reload`s to the assertion and it passed locally. Who knows what Travis will make of it. 😩

File: spec/controllers/admin/admin_users_controller_spec.rb [hunk 8]
```diff
@@ -92,24 +92,71 @@
...
+
+          it "allows admins with #{admin_role} role to update roles" do
+            put :update, params: { id: user.login, user: { roles: [role.id.to_s] } }
+            expect(user.reload.roles).to include(role)
+            it_redirects_to_with_notice(root_path, "User was successfully updated.")
+          end
+        end
+      end
+
+      %w[support policy_and_abuse].each do |admin_role|
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply spec/controllers/admin/admin_users_controller_spec.rb:135]: The Hound effing taught us to change to `%w[]`, and now it's changing its mind again? Not today, Hound.

File: app/models/comment.rb (hunk 2 above)
Comment (tickinginstant) [reply app/models/comment.rb:73]: I think that's an intermittent error in the existing code -- I'm pretty sure I ran into the same error in a Travis run for a completely unrelated branch, though unfortunately I restarted the the job (rather than preserving the log) when I realized it was intermittent, so I can't verify that it was the right intermittent failure.

If I'm not mistaken, the intermittent error occurs because if the test runs too quickly, the second comment edit doesn't change `edited_at`, which means that `saved_change_to_edited_at?` doesn't return true. That means that the comment isn't treated like a normal edited comment is (and therefore doesn't have its status set back to `unreviewed`).

I added some debugging code to print out `edited_at` and `edited_at_before_last_save`, and ran the test on this branch several times. The success/failure was determined in each case by whether the two `edited_at` values matched. Unfortunately, I'm having trouble replicating the same behavior on the master branch -- the test on master seems to be running much slower, so I have yet to be able to replicate the intermittent failure. (Though I guess that's good news for the upgrade, if it runs faster?)

File: app/models/tag.rb [hunk 9]
```diff
@@ -451,11 +451,11 @@ def self.in_prompt_restriction(restriction)
 
   def self.by_name_without_articles(fieldname = "name")
     fieldname = "name" unless fieldname.match(/^([\w]+\.)?[\w]+$/)
-    order("case when lower(substring(#{fieldname} from 1 for 4)) = 'the ' then substring(#{fieldname} from 5)
+    order(Arel.sql("case when lower(substring(#{fieldname} from 1 for 4)) = 'the ' then substring(#{fieldname} from 5)
             when lower(substring(#{fieldname} from 1 for 2)) = 'a ' then substring(#{fieldname} from 3)
             when lower(substring(#{fieldname} from 1 for 3)) = 'an ' then substring(#{fieldname} from 4)
             else #{fieldname}
-            end")
+            end"))
```
Comment (tickinginstant) [inline app/models/tag.rb:458]: Does `TagSetAssociation.by_name_without_articles` need similar treatment?
https://github.com/otwcode/otwarchive/blob/6dc754890a5cb9fb7da38a46fcf917998db3e63c/app/models/tagset_models/tag_set_association.rb#L15-L23
(Is raw SQL in `order` deprecated in 5.1 and needed to upgrade to 5.2, or is it deprecated in 5.2 and needed to upgrade to 6.0?)

Comment (CristinaRO) [reply app/models/tag.rb:458]: Possibly! I fixed them as I noticed them in the test suite output, so either I haven't noticed the one for `TagSetAssociation` or (frightening prospect) there is no test that exercises that method. (I prefer thinking that I haven't noticed that deprecation warning. 😓 )

Using "non-attribute arguments" in query methods[1] is a deprecation warning in Rails 5.2, and will become an exception in Rails 6.0. So it's not strictly necessary to upgrade to Rails 5.2, but I'm trying to put us ahead of the game (because that's what Elz was doing in her original branch 😅).

[1] The official verbiage is "Non-attribute arguments will be disallowed in Rails 6.0. This method should not be called with user-provided values, such as request parameters or model attributes. Known-safe values can be passed by wrapping them in Arel.sql()."

I still haven't fully digested that paragraph. If it's not best practice to use model attributes (because they can contain user input), why are they saying "non-attribute" arguments will be disallowed? Aren't model attributes "attribute arguments"? I find it very confusing.

Also, wrapping them in `Arel.sql()` isn't making anyone safer, since they say "known-safe values" - so it's still on us to be sure that those are safe values. (I think they are in this case, because we'll have done the sanitisation before persisting them.)

Comment (tickinginstant) [reply app/models/tag.rb:458]: It's not a good idea to use the values of model attributes unquoted, but the names of the attributes are generally fine. That's most likely what they mean by non-attribute arguments -- arguments that don't look like standard attribute names (aka column names).

**Comment by CristinaRO** [comment]
All deprecation warnings relating to Rails 6.0 are now gone, with one notable exception. (I haven't merged with the latest master though, and won't until this is getting tagged for a known release.)

The exception is
```
DEPRECATION WARNING: Using a dynamic :action segment in a route is deprecated and will be removed in Rails 6.0. (called from block in <top (required)> at otwarchive/config/routes.rb:619)
```
which I'm happy (you know, "happy") to leave for whomever we foist off the Rails 6.0 upgrade onto.

File: app/controllers/collection_items_controller.rb [hunk 10]
```diff
@@ -186,7 +190,16 @@ def update_multiple_with_params(allowed_items, update_params, success_path)
...
     # the user tries to update an item they're not allowed to.
     allowed_items.where(id: update_params.keys).each do |item|
-      @collection_items << item unless item.update(update_params[item.id])
+      item_data = update_params[item.id]
+      if item_data[:remove] == "1"
+        unless item.destroy
+          @collection_items << item
+        end
+      else
+        unless item.update(item_data)
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/controllers/collection_items_controller.rb:199]: Seems legit, will do in a future commit.

File: app/controllers/collection_items_controller.rb [hunk 11]
```diff
@@ -186,7 +190,16 @@ def update_multiple_with_params(allowed_items, update_params, success_path)
     # which uses find() under the hood -- we ensure that we'll fail silently if
     # the user tries to update an item they're not allowed to.
     allowed_items.where(id: update_params.keys).each do |item|
-      @collection_items << item unless item.update(update_params[item.id])
+      item_data = update_params[item.id]
+      if item_data[:remove] == "1"
+        unless item.destroy
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/controllers/collection_items_controller.rb:195]: Ditto

File: app/helpers/search_helper.rb [hunk 12]
```diff
@@ -2,41 +2,42 @@ module SearchHelper
...
+    case parent
+    when Collection
+      header << "in #{link_to(parent.title, parent)}"
+    when Pseud
+      header << "by #{parent.byline}"
+    when User
+      header << "by #{parent.login}"
+    end
+
+    if parent.is_a?(Tag)
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/helpers/search_helper.rb:32]: Will do in a separate commit, so it's easier to revert if people don't think it's actually better (more readable). ("Modifier" ifs are harder to spot at the end of longish lines, and make it look like the thing is being done - before I spot that it's a conditional.)

File: app/helpers/search_helper.rb [hunk 13]
```diff
@@ -2,41 +2,42 @@ module SearchHelper
...
-      header = ts "Recent #{item_name.pluralize}"
+      header << ts("Recent #{item_name.pluralize}")
     elsif collection.total_pages < 2
-      header = pluralize(collection.size, item_name)
+      header << pluralize(collection.size, item_name)
     else
       total_entries = collection.total_entries
       total_entries = collection.unlimited_total_entries if collection.respond_to?(:unlimited_total_entries)
-      header = %{ %d - %d of %d }% [
+      header << %{ %d - %d of %d }% [
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/helpers/search_helper.rb:13]: Uh, I have no idea what's going on here. Might research later, since "baffling" is not a very maintainable state for code.

File: app/models/collection.rb [hunk 14]
```diff
@@ -43,12 +43,12 @@ def ensure_associated
...
-      potential_matches.each {|potential_match| potential_match.destroy}
-      signups.each {|signup| signup.destroy}
-      prompts.each {|prompt| prompt.destroy}
-    end
+    return if self.challenge_id
+
+    assignments.each {|assignment| assignment.destroy}
+    potential_matches.each {|potential_match| potential_match.destroy}
+    signups.each {|signup| signup.destroy}
+    prompts.each {|prompt| prompt.destroy}
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/models/collection.rb:51]: Woof, woof.

File: app/helpers/search_helper.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/helpers/search_helper.rb:13]: Thinking out loud:

What do we gain here by quoting a string with the percent notation (`%{}`)? The fact that we then use the percent notation for string formatting (`%d`) and the operator `%`, for an amazing total of 3 different meanings for the same cursed symbol in one line of code, makes it a contender for obfuscated Ruby.

The cop isn't helping much by plonking all the "offenses" at once, even though the final suggestion seems to be "get rid of one of these `%`",  but at least it got me to understand the many meanings of `%`. 🤔

[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/helpers/search_helper.rb:13]: Do we need `format`, specifically, for its formatting properties? I assume we're using `%d` because we want to get something out of the number format. How would we express that as a template token like `%{foo}`? Is it `%d{foo}`?

Is this format compatible with `ts`, so we can translate the entire header?

File: app/helpers/search_helper.rb [hunk 15]
```diff
@@ -2,41 +2,34 @@ module SearchHelper
...
+      header << ts("in %{collection_link}", collection_link: link_to(parent.title, parent))
+    when Pseud
+      header << ts("by %{byline}", byline: parent.byline)
+    when User
+      header << ts("by %{username}", username: parent.login)
     end
-    header.html_safe
+
+    header << ts("in %{tag_link}", tag_link: link_to_tag_with_text(parent, parent.name)) if parent.is_a?(Tag)
+    header << ts("in %{fandom_link}", fandom_link: link_to_tag(@fandom)) if @fandom.present?
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/helpers/search_helper.rb:31]: I'd _love_ to, but I don't have enough context for how this method arrives at the point at having a `@fandom` instantiated in the first place, whether it's the same as the `parent` entity, and what we could use instead of an instance variable. :/

[the remark this answers is not in the document]
Comment (sarken) [reply app/helpers/search_helper.rb:31]: I have figured out where this is used! It's if you follow a fandom link on someone's dashboard, which will take you to a URL like /users/sarken/works?fandom_id=164274.

However, I don't think it's super necessary to do anything about this here.

File: spec/spec_helper.rb [hunk 16]
```diff
@@ -1,28 +1,29 @@
...
+require "email_spec"
+require "webmock/rspec"
 
 DatabaseCleaner.start
 DatabaseCleaner.clean
 
 # Requires supporting ruby files with custom matchers and macros, etc,
 # in spec/support/ and its subdirectories.
-Dir[Rails.root.join("spec/support/**/*.rb")].each {|f| require f}
+Dir[Rails.root.join("spec/support/**/*.rb")].sort.each { |f| require f }
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply spec/spec_helper.rb:23]: Because the Hound said so.

File: app/views/home/index.html.erb [hunk 17]
```diff
@@ -50,7 +50,7 @@
 
   <div class="latest tweets module">
     <h3 class="heading"><%= ts('Tweets') %></h3>
-    <a class="twitter-timeline" href="https://twitter.com/otw_status/lists/otw-tweets" data-widget-id="382983150978076672"><%= ts('Tweets from https://twitter.com/otw_status/lists/otw-tweets') %></a>
+    <a class="twitter-timeline" href="https://twitter.com/otw_status/lists/otw-tweets"><%= ts('Tweets from https://twitter.com/otw_status/lists/otw-tweets') %></a>
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/views/home/index.html.erb:53]: Yup, it makes it much easier to sift through the output when we don't have to scroll through screens of the same warning.

File: spec/spec_helper.rb (hunk 16 above)
[the remark this answers is not in the document]
Comment (redsummernight) [reply spec/spec_helper.rb:23]: For the record, the warning raised by Hound was [Lint/NonDeterministicRequireOrder](https://docs.rubocop.org/rubocop/cops_lint.html#lintnondeterministicrequireorder).

File: Gemfile.lock [hunk 18]
```diff
@@ -1035,7 +1035,9 @@ GEM
     chronic (0.10.2)
     climate_control (0.2.0)
     cliver (0.3.2)
-    codecov (0.1.20)
+    cocaine (0.5.8)
```
Comment (tickinginstant) [inline Gemfile.lock]: I think the merge broke this -- `cocaine` was renamed to `terrapin`, so I don't think it should be included in `Gemfile.lock`.

Comment (CristinaRO) [reply Gemfile.lock]: 😭 😭 😭 😭

**Comment by tickinginstant** [comment]
Well, that's interesting. Seems to be related to [this](https://github.com/rails/rails/issues/32870). Looks like the code for making sure the database starts with a default language/locale/admin setting is causing `rake db:create` to crash, because as of 5.2 the `db:create` task loads initializers. It wasn't a problem before the merge because the old `scripts/travis/mysql.sh` manually constructed the database, but #3876 removed that and switched over to using `rake db:create`.

I think if you rescue `ActiveRecord::NoDatabaseError` in the relevant config files, that should fix it. Though I don't know if that's the cleanest way to handle it.

File: app/models/comment.rb (hunk 2 above)
Comment (redsummernight) [reply app/models/comment.rb:73]: I got this error pretty consistently if I remove the `reload`s from the unreviewed checking steps, was this the one you ran into?

https://travis-ci.com/github/redsummernight/otwarchive/jobs/379025244

Comment (tickinginstant) [reply app/models/comment.rb:73]: No, that's a different intermittent error that's been happening on a fairly regular basis. That's spam vs. not spam (the `approved` column), not reviewed vs. unreviewed (the `unreviewed` column). Not sure why the `reload` would affect that, since the steps aren't used there. Probably just luck.

The only tests that use the two `marked as unreviewed` steps are in `comment_moderation.feature`. The intermittent failure I was talking about was on this test, IIRC:
https://github.com/otwcode/otwarchive/blob/b850d356002a00c4b57a679f64baf03662bf629a/features/comments_and_kudos/comment_moderation.feature#L262-L283

The added `reload` definitely doesn't help resolve the the issue with that test, since the step already loads a fresh copy of the comment even without calling `reload`. It might help with the one test where the step is called twice, by acting as a delay to help ensure that the timestamps change, but if we want a delay it's probably better to explicitly wait 1 second.

Comment (redsummernight) [reply app/models/comment.rb:73]: Gotcha, thanks! Took a few retries, but [this build](https://travis-ci.com/github/redsummernight/otwarchive/jobs/379025244) is passing now. I'll remove the `reload`s and we can add delays outside of this PR.
